ast/Trees.h

```diff
@@ -179,15 +183,19 @@ class TreePtr {
     }
 
     Expression *get() const noexcept {
-        auto val = reinterpret_cast<long long>(ptr) & PTR_MASK;
+        auto val = ptr & PTR_MASK;
 
-        // sign extension for the upper 16 bits
-        return reinterpret_cast<Expression *>((val << 16) >> 16);
+        if constexpr (sizeof(void *) == 4) {
```

Comment:
What's this guard here for? I don't see us checking sizeof in any other parts of the code, but maybe that's just because it doesn't show up in this diff?

Reply:
Ah I guess it's because in all other cases, even on 32-bit architectures, we'd still have 64 bits for tagged pointer storage, and this is the only interface where we expose a raw pointer?

Reply:
We don't need to clear the upper bits of the `long long` if pointers are only 4-bytes wide, but since we're already casting and throwing away the upper bits, this probably doesn't need to be here.

Reply:
And the reinterpret_cast is ok because on those architectures we'd be using only the bottom 32 bits, instead of the bottom 48 bits, and so the reinterpret cast does the same as what the 16 bit shifts do in the other case?